runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,601 @@
+        new DataSource<>(
+            context.getExecutionEnvironment(),
+            new ImpulseInputFormat(),
+            typeInformation,
+            pTransform.getUniqueName());
+
+    context.addDataSet(Iterables.getOnlyElement(pTransform.getOutputsMap().values()), dataSource);
+  }
+
+  /**
+   * Combiner that combines {@code T}s into a single {@code List<T>} containing all inputs.
+   *
+   * <p>For internal use to translate {@link GroupByKey}. For a large {@link PCollection} this is
+   * expected to crash!
+   *
+   * <p>This is copied from the dataflow runner code.
+   *
+   * @param <T> the type of elements to concatenate.
+   */
+  private static class Concatenate<T> extends Combine.CombineFn<T, List<T>, List<T>> {
+    @Override
+    public List<T> createAccumulator() {
+      return new ArrayList<>();
+    }
+
+    @Override
+    public List<T> addInput(List<T> accumulator, T input) {
+      accumulator.add(input);
+      return accumulator;
+    }
+
+    @Override
+    public List<T> mergeAccumulators(Iterable<List<T>> accumulators) {
+      List<T> result = createAccumulator();
+      for (List<T> accumulator : accumulators) {
+        result.addAll(accumulator);
+      }
+      return result;
+    }
+
+    @Override
+    public List<T> extractOutput(List<T> accumulator) {
+      return accumulator;
+    }
+
+    @Override
+    public Coder<List<T>> getAccumulatorCoder(CoderRegistry registry, Coder<T> inputCoder) {
+      return ListCoder.of(inputCoder);
+    }
+
+    @Override
+    public Coder<List<T>> getDefaultOutputCoder(CoderRegistry registry, Coder<T> inputCoder) {
+      return ListCoder.of(inputCoder);
+    }
+  }
+
+  private static void urnNotFound(
+      String id, RunnerApi.Pipeline pipeline, BatchTranslationContext context) {
+    throw new IllegalArgumentException(
+        String.format(
+            "Unknown type of URN %s for PTrasnform with id %s.",
+            pipeline.getComponents().getTransformsOrThrow(id).getSpec().getUrn(), id));
+  }
+
+  private static void pruneOutput(
+      DataSet<RawUnionValue> taggedDataset,
+      BatchTranslationContext context,
+      int unionTag,
+      Coder<WindowedValue<Object>> outputCoder,
+      String transformName,
+      String collectionId) {
+    TypeInformation<WindowedValue<Object>> outputType = new CoderTypeInformation<>(outputCoder);
+    FlinkExecutableStagePruningFunction<Object> pruningFunction =
+        new FlinkExecutableStagePruningFunction<>(unionTag);
+    FlatMapOperator<RawUnionValue, WindowedValue<Object>> pruningOperator =
+        new FlatMapOperator<>(
+            taggedDataset,
+            outputType,
+            pruningFunction,
+            String.format("%s/out.%d", transformName, unionTag));
+    context.addDataSet(collectionId, pruningOperator);
+  }
+
+  // Creates a mapping from PCollection id to output tag integer.
+  private static BiMap<String, Integer> createOutputMap(Iterable<String> localOutputs) {
```

Comment:
This and `instantiateCoder` are also used by the streaming translator. We can move them to a shared utility now or when we add the streaming translator?